### DISCUSSION TITLE
Why the component does not get re-rendered after value changed?

### DISCUSSION THREAD

**Discussion by softshipper** [pmndrs/jotai#964] [Q&A]
Hi all,
I am using Jotai in React Native for the first time in my life and I have an issue that the component does not get re-rendered after the value has changed. 
The atom definition:
```
export const contactsAtom: PrimitiveAtom<Contact[]> = atom<Contact[]>([]);
```
Here is how I use the `contactsAtom` atom for writing values in the `App.tsx` file:
```
const App: React.VFC = () => {
  const [loading, contacts, error] = useContacts();
  const setContacts = useUpdateAtom(contactsAtom);

  useEffect(() => {
    if (!loading) {
      console.log('set contact');
      setContacts(contacts);
    }
    console.log('app');
  }, [loading, contacts, setContacts]);

  if (loading) {
    return <Splash />;
  }

  if (error) {
    return <ErrorDesk />;
  }

  return (
    <NavigationContainer>
      <JotaiProvider>
        <Tabs />
      </JotaiProvider>
    </NavigationContainer>
  );
};
export default App;
```
As you can see the `contactsAtom` atom  is getting set with values when the `loading` variable equal `false` from the `useContacts` hooks. 

In an other component I read the value from the `contactsAtom` atom:
```
import {useAtomValue} from 'jotai/utils';
import React, {useEffect} from 'react';
import {Text, View} from 'react-native';
import {contactsAtom} from '../state';

export const Upcoming: React.VFC = () => {
  const contacts = useAtomValue(contactsAtom);
  useEffect(() => {
    console.log('upcoming');
  });
  return (
    <View>
      {contacts.map(c => (
        <Text>{c.givenName}</Text>
      ))}
    </View>
  );
};
```
Unfortunately, it does not show me a list of `givenName`. It seems that the component `Upcoming` does not get re-rendered when the `contactsAtom` atom  is getting set with values in the `App` component. 

What am I doing wrong?

As I have mentioned at the beginning, I am using Jotai in React Native and for app navigation https://reactnavigation.org/. 

Best regards,
Softshipper

**Comment by dai-shi** [comment]
```js
const setContacts = useUpdateAtom(contactsAtom);
```
☝️ That line is outside `<JotaiProvider>`. You need to use the hook under the provider tree.
An easy fix for your case is probably to remove `<JotaiProvider>` (which runs in provider-less mode.)

**Reply by softshipper** [reply]
First of all, thanks so much for your answer. It works like a charm. One more question, I have a custom hook and would like to know how to convert into atom?

````
import {useEffect, useState} from 'react';
import {getAll, Contact} from 'react-native-contacts';
import {Platform} from 'react-native';
import {PermissionsAndroid} from 'react-native';

export const useContacts = (): [boolean, Contact[], Error | undefined] => {
  const [loading, setLoading] = useState<boolean>(true);
  const [contacts, setContacts] = useState<Contact[]>([]);
  const [error, setError] = useState<Error | undefined>(undefined);

  const readContacts = (): Promise<void> =>
    getAll()
      .then(c => {
        setContacts(c);
        setLoading(false);
      })
      .catch(e => setError(e));

  useEffect(() => {
    if (Platform.OS === 'ios') {
      readContacts();
      return;
    }

    PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS, {
      title: 'Contacts',
      message: 'This app would like to view your contacts.',
      buttonPositive: 'Please accept bare mortal',
    }).then(readContacts);
  }, []);

  return [loading, contacts, error];
};
```
You have mentioned here https://github.com/pmndrs/jotai/discussions/896#discussioncomment-1830850 that custom hoook should be avoided right instead to atoms? 
Thanks

**Reply by dai-shi** [reply]
In a very jotai way, I wouldn't use useEffect, but sometimes useEffect is easy, especially converting from existing code.

To demonstrate how to do it with only an atom, here's a trial:
```js
const contactsAtom = atom(() => {
  if (Platform.OS === 'ios') {
    return getAll()
  }
  return PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS, {
    title: 'Contacts',
    message: 'This app would like to view your contacts.',
    buttonPositive: 'Please accept bare mortal',
  }).then(() => getAll())
})
```

**Reply by softshipper** [reply]
Could you please do the full example with `loading` and `error` variable like I do with custom hook. That's would be great. 
When the contact function needs time to load, I would show the `Loading` screen. If error occurs, then the `Error` screen.  

Thanks

**Reply by dai-shi** [reply]
There are no `loading` and `error` in this case. Just use `<Suspense fallback={<Loading />}>` and ErrorBoundary for errors. Unless you are familiar with them, it would not be very easy.
That's why I'm not super recommending it. Feel free to use custom hooks if you prefer.

**Reply by softshipper** [reply]
How to do it in the right way with Jotai? I am pretty sure that I will not the last ask this question. Please show us with an example, how to solve such as scenario with Jotai. That would be great. 

Best regards

**Reply by dai-shi** [reply]
Unfortunately, we don't have a great example to show what you really want. Need to improve docs.
Please check this for `<Suspense>` usage.
https://githubbox.com/pmndrs/jotai/tree/main/examples/hacker_news

For ErrorBoundary, there are no examples. Please check this package for the usage.
https://www.npmjs.com/package/react-error-boundary

> the right way with Jotai?

Please don't get it wrong. Using custom hooks is the right way too.
I'm not even sure how common Suspense is in RN development.
Maybe we should focus this discussion on custom hooks?

> I will not the last ask this question.

Yeah, we need better explanation. This is not strictly tied to Jotai, but we should show the usage clearly.
(If anyone is interested in helping docs, let us know.)

**Reply by softshipper** [reply]
I am interested to help but I have to get more into Jotai. 

It would be great, if you can show how to translate custom hooks into Jotai atoms in the documentation. 
Jotai is such a great library and I think it has a lot of potential. 
I am appreciate your work very much. 

Thanks a lot and have a greate day.

**Reply by dai-shi** [reply]
> how to translate custom hooks into Jotai atoms

This is not simple. It depends.

## useState

The basic one:
```js
const useFoo = () => {
  const [foo, setFoo] = useState('foo')
  // ...
}
```
can be translated into:
```js
const fooAtom = atom('foo')
const useFoo = () => {
  const [foo, setFoo] = useAtom(fooAtom)
  // ...
}
```

## useContacts

So, your case:
```js
const useContacts = () => {
  const [loading, setLoading] = useState(true)
  const [contacts, setContacts] = useState([])
  const [error, setError] = useState(undefined)

  const readContacts = () =>
    getAll()
      .then(c => {
        setContacts(c)
        setLoading(false)
      })
      .catch(e => setError(e))

  useEffect(() => {
    if (Platform.OS === 'ios') {
      readContacts()
      return
    }

    PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS, {
      title: 'Contacts',
      message: 'This app would like to view your contacts.',
      buttonPositive: 'Please accept bare mortal',
    }).then(readContacts)
  }, [])

  return [loading, contacts, error]
}
```
can be translated into:
```js
const contactsAtom = atom({
  loading: true,
  contacts: [],
  error: undefined,
}) // can be three atoms
const useContacts = () => {
  const [contacts, setContacts] = useAtom(contactsAtom)

  const readContacts = () =>
    getAll()
      .then(c => {
        setContacts({ contacts: c, loading: false })
      })
      .catch(e => setContacts({ error: e })

  useEffect(() => {
    if (Platform.OS === 'ios') {
      readContacts()
      return
    }

    PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.READ_CONTACTS, {
      title: 'Contacts',
      message: 'This app would like to view your contacts.',
      buttonPositive: 'Please accept bare mortal',
    }).then(readContacts)
  }, [])

  return contacts // or [contacts.loading, contacts.contacts, contacts.error]
}
```

**Reply by softshipper** [reply]
@dai-shi  thanks so much.

**Reply by dgknca** [reply]
the solution for me is to remove the Provider.
